### PR TITLE
Add experimental_implements attribute to dense_image_warp

### PR DESCRIPTION
**Pull request by thaink**
This attribute will help the TFLite conversion convert the whole block to a single TFLite custom op, significantly reduce the runtime.

# Description

Add experimental_implements attribute to dense_image_warp

## Type of change

- [ ] Bug fix
- [ ] New Tutorial
- [ ] Updated or additional documentation
- [ ] Additional Testing
- [ ] New Activation and the changes conform to the [activation contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/activations/README.md#contribution-guidelines)
- [ ] New Callback and the changes conform to the [callback contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/callbacks/README.md#contribution-guidelines)
- [x] New Image addition and the changes conform to the [image op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/image/README.md#contribution-guidelines)
- [ ] New Layer and the changes conform to the [layer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/layers/README.md#contribution-guidelines)
- [ ] New Loss and the changes conform to the [loss contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/losses/README.md#contribution-guidelines)
- [ ] New Metric and the changes conform to the [metric contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/metrics/README.md#contribution-guidelines)
- [ ] New Optimizer and the changes conform to the [optimizer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/optimizers/README.md#contribution-guidelines)
- [ ] New RNN Cell and the changes conform to the [rnn contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/rnn/README.md#contribution-guidelines)
- [ ] New Seq2seq addition and the changes conform to the [seq2seq contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/seq2seq/README.md#contribution-guidelines)
- [ ] New Text addition and the changes conform to the [text op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/text/README.md#contribution-guidelines)

# Checklist:

- [x] I've properly [formatted my code according to the guidelines](https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md#coding-style)
    - [x] By running Black + Flake8
    - [ ] By running pre-commit hooks
- [ ] This PR addresses an already submitted issue for TensorFlow Addons
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] This PR contains modifications to C++ custom-ops

# How Has This Been Tested?

### CODE REVIEW

**Comment by WindQAQ** [comment]
Hi @thaink, thanks for bringing this up. Seems that it breaks the backprop and unranked tensor inference, can you figure out why?

Also, it would be great to add yourself to CODEOWNER to get the notification when the implementation changes (In case you need to rewrite mlir patterns).

Thanks!

**Comment by thaink** [comment]
It is weird to break those as this PR only add a name to the tf.function.
Let me investigate why.

**Comment by thaink** [comment]
@WindQAQ I fixed the error by adding a new function. WDYT?

I am not really understand the code so well. So might not fit to put myself to CODEOWERS for this file.

**Review by WindQAQ — APPROVED** [review]
LGTM if we don't export it.

Is there any bug or somthing else in backprop? Will it hurt for any function annotated with `experimental_implements`, such as  `max_unpool2d`?

**Comment by thaink** [comment]
According to https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/eager/function.py#L1544, Function has attr cannot depend on any tensors outside of its signature or modify variables. But the backprop implicitly add some captured input to it.

@marksandler2 is my understand correct? Although the dense_image_warp function here does not use any outside variables but look like some are added under tf.GradientTape?

**Comment by marksandler2** [comment]
I think this is intended though unfortunate behavior. When a function is executed under gradient tape it now produces additional outputs that no longer have stable signature (e.g. they depend on the internal implementation of tf.function).
Thus it can't have the experimental_implement annotation. One option is to drop the annotation in those situations, which would make total sense in situations like you've encountered, however i am afraid that this would lead to silent dropping of annotations when the change is accidental. 

@thaink can you open a separate issue on Tensorflow side and cc me?

**Comment by marksandler2** [comment]
FYI: the issue has been fixed at tensorflow head

**Comment by thaink** [comment]
Hi @WindQAQ,
The since the fix is in tf-nightly, can I make a PR to merge dense_image_warp and dense_image_warp_annotated?

**Comment by thaink** [comment]
@WindQAQ In my test the test_unknown_shapes in tensorflow_addons/image/tests/dense_image_warp_test.py failed after the fix. Is it OK drop the unknown shape support?

@marksandler2 The error I mentioned above is at here:
fn = dense_image_warp.get_concrete_function(
        tf.TensorSpec(shape=None, dtype=image_type),
        tf.TensorSpec(shape=None, dtype=flow_type),
    )
Does adding experimental_implements change the behavior of get_concrete_function as well?

**Comment by marksandler2** [comment]
Thanks @Thai Nguyen <thaink@google.com>  - this looks unrelated to the
original issue. The problem seems to be in calling get_concrete_function
with TensorSpec rather than Tensor - i don't any documentation reflecting
this functionality.  It should however be easy to fix though - can you file
another bug?

**Comment by WindQAQ** [comment]
> Hi @WindQAQ,
> The since the fix is in tf-nightly, can I make a PR to merge dense_image_warp and dense_image_warp_annotated?

I'm afraid that our nightly version only pin to stable version of TensorFlow, say 2.3 and 2.4 now, so it's unlikely to merge it. I suggest that we can do that after TensorFlow 2.5 releases.

> @WindQAQ In my test the test_unknown_shapes in tensorflow_addons/image/tests/dense_image_warp_test.py failed after the fix. Is it OK drop the unknown shape support?

I think that test failed even before the fix? There should be another bug in TensorFlow. Unfortunately, we cannot drop it as some `tf.data.Dataset` pipeline will generate unranked tensor.

> The problem seems to be in calling get_concrete_function with TensorSpec rather than Tensor - i don't any documentation reflecting this functionality.

The functionality is documented here.

https://www.tensorflow.org/guide/function#obtaining_concrete_functions

**Comment by marksandler2** [comment]
Yeah - that's an unrelated bug! There were actually several other
alternative codepaths broken with
TensorSpec provided instead of a tensor. Will follow up with tensorflow.

**Comment by thaink** [comment]
Thanks Mark for the quick investigation.
@WindQAQ I'll make a PR saying dense_image_warp_annotated will be removed.
